### ISSUE TITLE
conflicts between services using ports and containers using ports

### ISSUE THREAD

**Issue by vikstrous** [moby/moby#34561]
**Description**

If a container uses a port and then a service is created with that port as a published port, bad things happen. The same issue happens in the other order as well.

**Steps to reproduce the issue:**
1. `docker service create --name testservice -p 2000:2000 alpine sleep 10000`
2. `docker run --rm -it -p 2000:2000 alpine sh`
`docker: Error response from daemon: driver failed programming external connectivity on endpoint flamboyant_yonath (08e19c9f4bc0b0429e37af9508102eb5bcc2975d3a906e959f56c09d8476125e): Error starting userland proxy: listen tcp 0.0.0.0:2000: bind: address already in use.`
3. `docker service rm testservice`
4. `docker run --rm -it -p 2000:2000 alpine sh`
`docker: Error response from daemon: driver failed programming external connectivity on endpoint amazing_knuth (66e2b938359c9e6cf7be433727c86676eef56add1d5fde33ac345005bc915db1): Error starting userland proxy: listen tcp 0.0.0.0:2000: bind: address already in use.`

1. `docker run --rm -it -p 2002:80 nginx:alpine`
2. `curl http://127.0.0.1:2002` -> works
3. `docker service create --name testservice -p 2002:2002 alpine sleep 10000`
4. `docker service ps testservice`
```
ID                  NAME                IMAGE               NODE                DESIRED STATE       CURRENT STATE            ERROR               PORTS
a03ed5uagf4z        testservice.1       alpine:latest       wonderwoman         Running             Running 14 seconds ago    
```
5. `curl http://127.0.0.1:2002` -> The port seems to be hijacked by the service at this point (connection refused)
6. `docker service rm testservice`
7. `curl http://127.0.0.1:2002` -> still connection refused
8. (kill container)
9. start container again
10. `curl http://127.0.0.1:2002` -> works

It looks like I need to wait some amount of time after the service is removed until the container can take back the port. I'm not entirely sure what's going on here.

**Describe the results you received:**
TLDR: ports being stolen, containers permanently not starting, etc.

**Describe the results you expected:**
Basically docker not being so confused about its state. Either containers take precedence over services or the other way around. I would prefer it if containers take precedence because the service port publication feature is a very big hammer. It takes up a port on every machine in the cluster and I don't think that's what most users want anyway. If containers take precedence, that gives you a bit more control to replace/override the service's claim.

**Additional information you deem important (e.g. issue happens only occasionally):**
Present in every docker/swarm version afaik, but probably with different symptoms.

**Output of `docker version`:**

```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:15:15 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:51:55 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

**Output of `docker info`:**

```
Containers: 21
 Running: 19
 Paused: 0
 Stopped: 2
Images: 2103
Server Version: 17.06.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: x7fwrj2eylvdng6voln3gmhgm
 Is Manager: true
 ClusterID: 63hxmw8i9cqu02q6eyj0exsog
 Managers: 1
 Nodes: 1
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
  External CAs:
    cfssl: https://172.17.0.1:12381/api/v1/cfssl/sign
 Root Rotation In Progress: false
 Node Address: 172.17.0.1
 Manager Addresses:
  172.17.0.1:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.12.4-1-ARCH
Operating System: Arch Linux
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.28GiB
Name: wonderwoman
ID: X6WI:7T7A:7GO5:ENSY:WKUY:AIKK:YLJE:WPNU:Z27O:O25G:7IHT:W3Q6
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 191
 Goroutines: 337
 System Time: 2017-08-18T13:25:27.728881472-07:00
 EventsListeners: 2
Username: viktorstanchev
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 172.17.0.1
 172.17.0.1:443
 172.17.0.1:445
 52.32.51.132:444
 172.17.0.1:3001
 172.17.0.1:444
 34.210.233.10
 35.167.19.104:444
 127.0.0.0/8
Live Restore Enabled: false

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical

**Comment by friism** [comment]
cc @mavenugo
